### ISSUE TITLE
VS: Create tool to generate VS .sln from .builds files

### ISSUE THREAD

**Issue by clrjunkie** [dotnet/corefx#4919]
Continuing #4901

@weshaggard

I suggest keeping this open just for a little bit longer.

>  We want to keep libraries isolated so they can ship independent of each other and in general we want the libraries to depend upon the lowest version of the package that contains the necessary dependencies, and not have project references to the live versions of their dependent libraries. 

This goal is completely orthogonal to the pain points I listed. Inter-project dependencies are a fact  and attempting to solve the granularity problem (Packages or not) by trading in developer productivity is simply a mistake, and if you insist  to compare apples to oranges than aiding to code comprehension is by far more important to correctness.  Technically there is nothing that prohibits someone from directly referencing any other project (csprog) . (something which is already being done between existing projects in the corefx repo). 

What constitutes the right level of package isolation can only be determined by peer code-review not the build system. Suggesting that independent  build scripts solve the problem is a misunderstanding of the problem because build scripts are also modified to create dependencies just in a less visible way.  

> For cases were folks have a desire to run tests against the live versions of all the dependencies built in the repo people can pass /p:TestWithLocalLibraries=true when building test projects

Building unit tests is a minor problem and has nothing to do with this issue.

The fundamental problem is Navigating the source (Goto To Definition)  and Debugging (Step Into) _across different implementation projects and test projects using Visual Studio_ again, aiding to code comprehension should be a key requirement.  (WinDbg - Forget it!)

We need to automatically generate a solution that references projects (csproj) that live in parallel to the independent existing project csproj. What gets into final Nuget packages or specified in the project.json is a completely different issue.

Failing to address this is a pushback to potential contributing (a.k.a volunteering) efforts - a financial lose for the project.

**Comment by Maxwe11** [comment]
Keeping ~100 projects in one solution it's a pain. I want to load project on my laptop, not mainframe.

**Comment by clrjunkie** [comment]
if you dont need too don't open the *corfx* solution file, open the projects' file, this about debugging/navigating the corefx not a project built to use corefx.

**Comment by mellinoe** [comment]
I also think the number of projects in here is probably too unwieldy to put into a single solution, especially because we just added a huge number of extra ones.

That said, it would be interesting if there was a tool that could give you a solution containing a custom subset of the libraries depending on what you wanted to debug. For example, maybe I could generate a solution containing all of the Threading, Task, and IO libraries to debug something I have, and leave all the rest out to reduce clutter.

**Comment by clrjunkie** [comment]
that's also a possibility

**Comment by weshaggard** [comment]
@clrjunkie sorry about cutting the other issue off my intention was not about trying to stop the conversation, you certainly could have continued the discussion over there. 

While I do understand the desire and I agree it would be nice to have the ability to jump around to the definition of all code throughout the entire repo, however I don't think we currently have the tools to make that work well. 

As has already been said maintaining a solution file for this entire repo would be a nightmare and even if we did keep up the solution we cannot change all the projects to be project references to each other as that would break our other goals of minimizing the dependencies to the lowest versions. One of the other challenges we have is all the various configurations for our libraries and so in order to truly navigate to all the correct code you would also need to know the correct configuration combinations within the solution for what you want to target (see https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/project-guidelines.md for an idea of all the different pivots). It is for these reasons I'm not sure the value is there for trying to maintain a solution for the entire repo, however if you are interested in trying to create a tool that would allow people to generate a solution on the fly that would be a great addition.

As for source navigation it is a goal of mine to eventually turn on indexing (think http://referencesource.microsoft.com/)  across this (and perhaps other .NET Core) repos that gets updated as part of our CI. That will allow for easier browsing and navigating of the code from the web browser. 

As for debugging in VS I think we mostly have that covered with the options available such that you can actually step into and through the code, if you do the debugging of one of our test projects with the `TestWithLocalLibraries` option. If you feel there is a particular scenario that doesn't work for you please pass that along and we will see what we can do about addressing it. 

>Technically there is nothing that prohibits someone from directly referencing any other project (csprog) . (something which is already being done between existing projects in the corefx repo).

Our current guidance is to only allow project references from test projects to the library that they are actually testing. There are some temporary exceptions to this rule but those are only exceptions and not the rule.

**Comment by clrjunkie** [comment]
@weshaggard

There are multiple concerns here that need to be discussed separately:

* VS solution with many CoreFx projects. 

I don't fully understand the concern and why this should be of a concern. Are we talking about the solution load time ? the compile time ? 

The corefx repo contains  478 csproj files and 7026 c#  files which is approximately 15 files per project.

Educated guess: The average developer uses a 64bit computer having  8 x ~2.3 GHz  cores, 8GB of ram and an SSD disk throughput of 500MB/s.  Fact: 40 years ago someone built a 64 bit computer named Cray-1 that was used for things like weather forecasting and nuclear engineering, which had  80Mhz CPU,  8MB of ram and 5MB /s disk throughput.

My point being: should there really be a performance problem? 

I started to hack a VS extension that will add all corefx projects to one solution but I decided to put in on hold until I fully understand the next point.

* Project version management

How are you tracking changes to individual projects by release/fix version? How can you build a  particular package version directly from the source repo?  Since each project is required to have its own version it seems like Git Submodules are much more appropriate for CoreFx than managing all the projects in one repository. 

* Library configuration (Pivot)

I only need to  select a particular VS build configuration  that applies to all corefx projects. VS will enable and disable all the relevant ifdefs. Where is the problem? 

* Source navigation using VS

I totally disagree this is just "nice". There is a reason why VS2015 introduced the new "Go To Implementation". feature.  It's much simpler and quicker to follow the source without needing to setup dummy test projects and breakpoints.

* Debugging using VS

So here is the scenario: I ship a product depending on corefx packages at specific version. There is a bug and I suspect it relates to a particular corefx package. I want to debug that package by stepping into its source code. Where can I find the matching source code?  (The repo might already contain a modified version).

**Comment by karelz** [comment]
Next steps: We need a tool to generate .sln files from existing .builds files.
Tricky question afterwards: Do we keep .sln checked in? If yes, how do we update them? (enforcement / ad-hoc)

**Comment by weshaggard** [comment]
We no longer have .builds files but we do have tooling to generate solution files per library, which is the level we think is the most useful for library developers. I'm closing this issue but if there is something we thinking we are still missing we can open another specific issue.
